src/compute-client/src/controller/replicated.rs

```diff
@@ -145,19 +146,109 @@ struct PendingPeek {
     otel_ctx: OpenTelemetryContext,
 }
 
+/// Reported upper frontiers for a single compute collection.
+#[derive(Debug)]
+struct ReportedUppers<T> {
+    /// The reported uppers per replica.
+    per_replica: HashMap<ReplicaId, Antichain<T>>,
+    /// The lower and upper bound of all reported uppers.
+    bounds: FrontierBounds<T>,
+}
+
+impl<T> ReportedUppers<T>
+where
+    T: Timestamp + Lattice,
+{
+    /// Construct a [`ReportedUppers`] that tracks frontiers of the given replicas.
+    fn new(replica_ids: &BTreeSet<ReplicaId>) -> Self {
+        let per_replica = replica_ids
+            .iter()
+            .map(|id| (*id, Antichain::from_elem(T::minimum())))
+            .collect();
+
+        Self {
+            per_replica,
+            bounds: FrontierBounds {
+                lower: Antichain::from_elem(T::minimum()),
+                upper: Antichain::from_elem(T::minimum()),
+            },
+        }
+    }
+
+    /// Start tracking the given replica.
+    ///
+    /// # Panics
+    /// - If the given `replica_id` is already tracked.
+    fn add_replica(&mut self, id: ReplicaId) {
+        let previous = self.per_replica.insert(id, self.bounds.lower.clone());
+        assert!(previous.is_none(), "replica already tracked");
+    }
+
+    /// Stop tracking the given replica.
+    ///
+    /// # Panics
+    /// - If the given `replica_id` is not tracked.
+    fn remove_replica(&mut self, id: ReplicaId) {
+        self.per_replica.remove(&id).expect("replica not tracked");
+    }
```

Comment:
Should these two update `self.bounds` at the same time? It seems like we risk leaving in a state that does not maintain invariants about `per_replica` and `bounds`.

Reply:
`add_replica` won't change the bounds, as we are just adding the current lower bound again. `remove_replica` can change the lower bound, if we remove the slowest replica. We could update here, but it's a bit tricky: If we update the lower bound, then we should also produce a `FrontierUppers` response to the compute controller. Which is information that we could propagate up, but it makes the control flow more complex.

So instead my thinking was that only `update` should change the bounds, so only callers of `update` need to worry about sending `FrontierUppers`. When you remove a replica, nothing will happen immediately. The next time we receive a `FrontierUppers` response from another replica, `update` will be called again and the new lower bound is reported.

But I'm now realizing that this doesn't work correctly when we remove the last replica, as we won't see any more `FrontierUppers` responses after that. So we need to produce bound updates in response to `remove_replica` after all!